### PR TITLE
V1 subscriptions 

### PR DESCRIPTION
**Pull request by fowbi**
Documentation for the Subscription endpoints

### CODE REVIEW

File: apiary.apib [hunk 1]
```diff
@@ -879,6 +879,284 @@ Delete a file
...
 
+## Subscriptions [/subscriptions]
+
+### /addSubscription.php [POST /addSubscription.php]
+
+Add subscription
+
++ Request (application/json)
+
+    Item lines can be added by giving the item line <id> as prefix. eg description_1 will contain the description for Item line 1, description_2 will contain the description for item line 2, ...
```
Comment (Brammm) [inline apiary.apib]: Q: suffix, not prefix, no?

Comment (Steerlin) [inline apiary.apib]: yes, should be suffix

File: apiary.apib [hunk 2]
```diff
@@ -879,6 +879,284 @@ Delete a file
...
+
+    + Attributes(object)
+        + name: `Johan` (string, required) - The name of the product
```
Comment (Steerlin) [inline apiary.apib]: if name is the name of the product, then 'Johan' is a very poor and confusing example. 
Better would be 'Teamleader CRM' or 'Acme SaaS Product'

Comment (Steerlin) [inline apiary.apib]: otherwise the description is wrong

File: apiary.apib [hunk 3]
```diff
@@ -879,6 +879,284 @@ Delete a file
...
+        + date_start: `01/08/1987` (string, required) - Date the first invoice should be created. The date should be in the format dd/mm/yyyy
+        + repeat_after: `monthly` (enum[string], required) - Repeat cycle 
+            + Members
+                + monthly
+                + twomonthly
+                + quarterly
+                + sixmonthly
+                + yearly
+                + twoyearly
+        + title: `example.com` (string, required) - eg: Hosting, Domain names,...
```
Comment (Steerlin) [inline apiary.apib]: also here - poor and confusing example data

File: apiary.apib [hunk 4]
```diff
@@ -879,6 +879,284 @@ Delete a file
...
+            + 10D
+            + 15D
+            + 21D
+            + 30D
+            + 45D
+            + 60D
+            + 30DEM
+            + 60DEM
+            + 90DEM
+        + custom_field_ID: `bar` (string, optional) - Replace ID by the ID of your custom field. (eg custom_field_310)
```
Comment (Steerlin) [inline apiary.apib]: here we use 'replace `_ID` ' vs `_<id>` on the next line

Comment (Steerlin) [inline apiary.apib]: in other PR's `_ID`seemed to be the standard ...

File: apiary.apib [hunk 5]
```diff
@@ -879,6 +879,284 @@ Delete a file
...
+            + Members
+                + 00
+                + 06
+                + 12
+                + 21
+                + CM
+                + EX
+                + MC
+                + VCMD
+        + product_<id>_1: `1` (number, optional) Id of the product for item line <id>
```
Comment (Steerlin) [inline apiary.apib]: missing '-' before field description

File: apiary.apib [hunk 6]
```diff
@@ -879,6 +879,284 @@ Delete a file
...
+        + account_<id>: `1` (number, optional)  Id of the bookkeeping account for item line <id>
```
Comment (Steerlin) [inline apiary.apib]: missing -

File: apiary.apib [hunk 7]
```diff
@@ -879,6 +879,284 @@ Delete a file
...
+        + subtitle_<id>: `Sub sub title` (string, optional) Subtitle for item line <id>
```
Comment (Steerlin) [inline apiary.apib]: missing -

File: apiary.apib [hunk 8]
```diff
@@ -879,6 +879,284 @@ Delete a file
...
+
++ Response 200 (application/json)
+    + Attributes (object)
+        + id: `416654` (number) - ID of the subscription that was added
```
Comment (Steerlin) [inline apiary.apib]: should we also include `status: 'success' ` so clients can generically check the responses for this OK-status indicator ?

File: apiary.apib [hunk 9]
```diff
@@ -879,6 +879,284 @@ Delete a file
...
+
+    + Attributes (array)
+        + (object)
+            + id: 48100 (number) - The id of the subscription
+            + active: 1 (enum[number]) - Boolean indicating if the invoice is active
+                + Members
+                    + 1
+                    + 0
+            + repeat: quarterly (enum[string])
+                + Members
```
Comment (Steerlin) [inline apiary.apib]: more options here than for getSubscription ?

File: apiary.apib [hunk 10]
```diff
@@ -879,6 +879,284 @@ Delete a file
...
+### /updateSubscription.php [POST /updateSubscription.php]
+
+    Update subscription
+
++ Request (application/json)
+
+    + Attributes
+        + subscription_id: `1` (number, required) - ID of the subscription
+        + date_start: `01/08/1987` (string, optional) - Date the first invoice should be created. The date should be in the format dd/mm/yyyy.
+        + repeat_after: `monthly` (enum[string], optional) -  Invoice repeat cycle
```
Comment (Steerlin) [inline apiary.apib]: less options here than getSubscription ?

File: apiary.apib [hunk 11]
```diff
@@ -879,6 +879,284 @@ Delete a file
...
+                    + 4yearly
+                    + 5yearly
+                    + twoweekly
+                    + 6yearly
+                    + 7yearly
+                    + 8yearly
+                    + 9yearly
+                    + 10yearly
+                    + fourmonthly
+            + client_name: Erlich Bachman (string) - Name of the client
```
Comment (Steerlin) [inline apiary.apib]: back-ticks seem to have disappeared buy now ... e.g. ``Erlich Bachman``

File: apiary.apib [hunk 12]
```diff
@@ -879,6 +879,284 @@ Delete a file
...
+            + department_name: Finance (string) - Department for which the subscription is created
+            + department_id: 2 (number) - Id of the department
+            + date_start: 1480287600 (number) - Unix timestamp of the start date of the subscription
+            + date_start_formatted: 28/11/2016 (string) - Formatted start date of the subscription
+            + date_end (number, nullable) - Unix timestamp of the end date of the subscription (can be empty)
```
Comment (Steerlin) [inline apiary.apib]: no example data provided

File: apiary.apib [hunk 13]
```diff
@@ -879,6 +879,284 @@ Delete a file
...
+            + date_end_formatted (string, nullable) Formatted end date of the subscription
```
Comment (Steerlin) [inline apiary.apib]: no example data provided

File: apiary.apib [hunk 14]
```diff
@@ -879,6 +879,284 @@ Delete a file
...
+            + contact_or_company: contact (enum[string]) - Indicates the subscription is linked to a contact or a company
+                + Members
+                    + contact
+                    + company
+            + contact_or_company_id: 1174438 (number) - The id of the contact or company
+            
+
+### /getInvoicesBySubscription.php [POST /getInvoicesBySubscription.php]
+
+    Fetching all invoices created by a subscription
```
Comment (Steerlin) [inline apiary.apib]: Fetching the id's of all invoices created by a subscription

File: apiary.apib [hunk 15]
```diff
@@ -879,6 +879,284 @@ Delete a file
...
+        + subscription_id: 1245 (number, required) - The id of the subscription
+
++ Response 200 (application/json)
+
+    + Attributes(object)
+        + generated_invoice_ids: 11400, 11401 (array[number])
+
+### /getRelatedSubscriptionsByInvoice.php [POST /getRelatedSubscriptionsByInvoice.php]
+
+    Fetching all subscriptions related to an invoice
```
Comment (Steerlin) [inline apiary.apib]: Fetching the id's of all subscriptions related to an invoice

File: apiary.apib [hunk 16]
```diff
@@ -879,6 +879,284 @@ Delete a file
...
++ Response 200 (application/json)
+
+    + Attributes(array)
+        + (object)
+            + subscription_id: 7457 (number)
+            + subscription_title: Monthly subscription Teamleader (string)
+
+### /getSubscriptionsByDeal.php [POST /getSubscriptionsByDeal.php]
+
+    Fetching all subscriptions related to a deal
```
Comment (Steerlin) [inline apiary.apib]: Fetching the id's of all subscriptions related to a deal

File: apiary.apib [hunk 17]
```diff
@@ -879,6 +879,284 @@ Delete a file
...
+
++ Request (application/json)
+
+    + Attributes(object)
+        + deal_id: 1337 (number, required) - The id of the deal
+
++ Response 200 (application/json)
+
+    + Attributes(object)
+        + related_subscription_ids: 1245, 2584 (array[number])
```
Comment (Steerlin) [inline apiary.apib]: seems comma separate (string) list in the example. Json array is `[1245, 2584]`

File: apiary.apib [hunk 18]
```diff
@@ -879,6 +879,313 @@ Delete a file
...
+                + 6yearly
+                + 7yearly
+                + 8yearly
+                + 9yearly
+                + 10yearly
+                + fourmonthly
+        + title: `Monthly subscription Teamleader` (string, required) - eg: Hosting, Domain names,...
+        + date_end: `01/08/1987` (string, optional) - Date this subscription should expire. Date should be in the format dd/mm/yyyy
+        + related_deal_id: `1` (number, optional) - ID of the deal this subscription is related to
+        + permanent_invoicing: `0` (string, optional) - When set to 0 this subscription won't be automatically booked. Default enabled (1).
```
Comment (jenssegers) [inline apiary.apib]: number or enum[number]

File: apiary.apib [hunk 19]
```diff
@@ -879,6 +879,313 @@ Delete a file
...
+        + autosend_invoice: `0` (string, optional) - When set to 1, this subscription invoice will be send automatically.
```
Comment (jenssegers) [inline apiary.apib]: number or enum[number]

File: apiary.apib [hunk 20]
```diff
@@ -879,6 +879,313 @@ Delete a file
...
+        + autosend_bpost: `0` (string, optional) - When set to 1, this subscription invoice through physical mail will be send automatically.
```
Comment (jenssegers) [inline apiary.apib]: number or enum[number]

File: apiary.apib [hunk 21]
```diff
@@ -879,6 +879,313 @@ Delete a file
...
+        + comments: `Comment` (string, optional) - Comments
+        + days_in_advance: `1` (string, optional) - Sending of the invoice. Set to 0 to send out the invoice on the day it's due. Defaults to 7 days
```
Comment (jenssegers) [inline apiary.apib]: number?

File: apiary.apib [hunk 22]
```diff
@@ -879,6 +879,313 @@ Delete a file
...
+            + 21D
+            + 30D
+            + 45D
+            + 60D
+            + 30DEM
+            + 60DEM
+            + 90DEM
+        + custom_field_ID: `bar` (string, optional) - Replace ID by the ID of your custom field. (eg custom_field_310)
+        + description_ID: `Item description` (string, optional) - Description of item line ID
+        + price_ID: `12.34` (string, optional) - Item line ID price
```
Comment (jenssegers) [inline apiary.apib]: number?

File: apiary.apib [hunk 23]
```diff
@@ -879,6 +879,313 @@ Delete a file
...
+        + amount_ID: `100` (string, optional) - Item line ID amount
```
Comment (jenssegers) [inline apiary.apib]: number?

File: apiary.apib [hunk 24]
```diff
@@ -879,6 +879,313 @@ Delete a file
...
+                + 8yearly
+                + 9yearly
+                + 10yearly
+                + fourmonthly
+        + department_id: `1` (number) - ID of the department
+        + date_start: `554774400` (number) - Start date of subscription as unix timestamp
+        + date_start_formatted: `01/08/1987` (string) - Start date of subscritpion in formatted string mm/dd/yyyy.
+        + date_end: `554774400` (number) -  End date of subscription as unix timestamp
+        + date_end_formatted: `01/08/1987` (string) - End date of subscritpion in formatted string mm/dd/yyyy.
+        + total_price_excl_vat: `12.34` (string) - Price excl VAT
```
Comment (jenssegers) [inline apiary.apib]: number?

File: apiary.apib [hunk 25]
```diff
@@ -879,6 +879,313 @@ Delete a file
...
+        + total_price_incl_vat: `12.34` (string) - Price incl VAT
```
Comment (jenssegers) [inline apiary.apib]: number?

File: apiary.apib [hunk 26]
```diff
@@ -879,6 +879,313 @@ Delete a file
...
+        + next_renewal_date: `554774400` (number) - Next renewal date of subscription as unix timestamp
+        + next_renewal_date_formatted: `01/08/1987` - (string) Next renewal date of subscritpion in formatted string mm/dd/yyyy.
+        + items: [] (array[object]) - Item line
```
Comment (jenssegers) [inline apiary.apib]: Did you test this in apiary? Not sure if this works?

File: apiary.apib [hunk 27]
```diff
@@ -879,6 +879,313 @@ Delete a file
...
+            + (object)
+                + text: `title` (string) - Title
+                + amount: `1` (number) - Item amount
+                + line_total_incl_vat `12.34` (string) - Line incl VAT
+                + line_total_excl_vat `12.34` (string) - Line excl VAT
```
Comment (jenssegers) [inline apiary.apib]: number?

File: apiary.apib [hunk 28]
```diff
@@ -879,6 +879,313 @@ Delete a file
...
+        + date_end_formatted: `01/08/1987` (string) - End date of subscritpion in formatted string mm/dd/yyyy.
+        + total_price_excl_vat: `12.34` (string) - Price excl VAT
+        + total_price_incl_vat: `12.34` (string) - Price incl VAT
+        + next_renewal_date: `554774400` (number) - Next renewal date of subscription as unix timestamp
+        + next_renewal_date_formatted: `01/08/1987` - (string) Next renewal date of subscritpion in formatted string mm/dd/yyyy.
+        + items: [] (array[object]) - Item line
+            + (object)
+                + text: `title` (string) - Title
+                + amount: `1` (number) - Item amount
+                + line_total_incl_vat `12.34` (string) - Line incl VAT
```
Comment (jenssegers) [inline apiary.apib]: number?

File: apiary.apib [hunk 29]
```diff
@@ -879,6 +879,313 @@ Delete a file
...
+                + line_total_excl_vat `12.34` (string) - Line excl VAT
+                + vat_rate `21` (number) - VAT rate
```
Comment (jenssegers) [inline apiary.apib]: string?

File: apiary.apib [hunk 30]
```diff
@@ -879,6 +879,313 @@ Delete a file
...
+### /getSubscriptions.php [POST /getSubscriptions.php]
+
+    Search subscriptions
+
++ Request (application/json)
+
+    + Attributes(object)
+        + amount: `50` (number, required) - The amount of subscriptions returned per request (between 1 and 100)
+        + pageno: `0` (number, required) - The page number (the first page is 0)
+        + selected_customfields: `14,15` (array[number]) - Comma-separated list of the IDs of the custom fields you wish to select (max 10)
```
Comment (jenssegers) [inline apiary.apib]: I think this needs to be without backticks.

File: apiary.apib [hunk 31]
```diff
@@ -879,6 +879,313 @@ Delete a file
...
+            + department_name: `Finance` (string) - Department for which the subscription is created
+            + department_id: `2` (number) - Id of the department
+            + date_start: `1480287600` (number) - Unix timestamp of the start date of the subscription
+            + date_start_formatted: `28/11/2016` (string) - Formatted start date of the subscription
+            + date_end `1480287600` (number, nullable) - Unix timestamp of the end date of the subscription (can be empty)
+            + date_end_formatted `28/11/2016` (string, nullable) - Formatted end date of the subscription
+            + total_price_excl_vat: `100.00` (number) - Price, excluding VAT
+            + total_price_incl_vat: `106.00` (number) - Price, including VAT
+            + next_renewal_date: `1488236400` (number) - Unix timestamp of the next renewal date
+            + next_renewal_date_formatted: `28/02/2017` - Formatted next renewal date
```
Comment (jenssegers) [inline apiary.apib]: Missing type

File: apiary.apib [hunk 32]
```diff
@@ -879,6 +879,313 @@ Delete a file
...
+
++ Request (application/json)
+
+    + Attributes(object)
+        + subscription_id: `1245` (number, required) - The id of the subscription
+
++ Response 200 (application/json)
+
+    + Attributes(object)
+        + generated_invoice_ids: `11400, 11401` (array[number])
```
Comment (jenssegers) [inline apiary.apib]: Also without backticks here I think

File: apiary.apib [hunk 33]
```diff
@@ -879,6 +879,313 @@ Delete a file
...
+
++ Request (application/json)
+
+    + Attributes(object)
+        + deal_id: `1337` (number, required) - The id of the deal
+
++ Response 200 (application/json)
+
+    + Attributes(object)
+        + related_subscription_ids: `1245, 2584` (array[number])
```
Comment (jenssegers) [inline apiary.apib]: Same here

**Pull request merged**